### ISSUE TITLE
Renders new line on every render if line width > terminal width

### ISSUE THREAD

**Issue by jsejcksn** [deno-library/progress#8]
Hello. I have noticed a bug, which is reproducible by running [the "simple" example in the readme](https://github.com/deno-library/progress/blob/v1.2.3/README.md#single-progress-bar) using the default terminal profile on macOS.

Here are the steps to reproduce:

<details>
<summary>Versions:</summary>

| type | name | version |
| :-- | :--- | :--- |
| os | macOS | `11.3.1 (20E241)` |
| terminal emulator | Terminal.app | `2.11 (440)` |
| runtime | Deno | `1.9.2` |
| module | progress | `1.2.3` |
</details>

1. Create a new **unmodified** profile in macOS Terminal:

    <img width="667" alt="0" src="https://user-images.githubusercontent.com/868251/117551794-996c3100-b00d-11eb-9d41-cb6d316f2a2a.png">

2. Open a new Terminal window using that profile:

    <img width="702" alt="1" src="https://user-images.githubusercontent.com/868251/117551807-a852e380-b00d-11eb-9364-8be18950de4f.png">

    <img width="570" alt="2" src="https://user-images.githubusercontent.com/868251/117551816-b0ab1e80-b00d-11eb-9d5c-e48944568726.png">

3. Create an example TypeScript module with the code from [the "simple" example in the readme](https://github.com/deno-library/progress/blob/v1.2.3/README.md#single-progress-bar):

    <img width="570" alt="3" src="https://user-images.githubusercontent.com/868251/117551843-d7695500-b00d-11eb-82c2-a6e4d2233b67.png">

4. Run the example file:

    https://user-images.githubusercontent.com/868251/117553293-624e4d80-b016-11eb-94f6-d1cfad95851a.mp4

---

As you can see in the video, it re-renders to the same line until the "completed" value reaches `10`, at which point the render string width exceeds the terminal width and each render begins to print to a new line.

> I think this is not a good experience (especially for users who are just trying the module).

If each progress bar can render more than one line, then I think this is simply a bug. However, I haven't researched the source of this module yet, but it seems like it only renders a single line for each progress bar. With that assumption in mind, here is an idea for how to address it:

- Add a new constructor option `maxWidth` to manually set the max length of each line rendered. By default, this value would be `Infinity` if left undefined.
- When a render occurs, the line is truncated to the length of the `maxWidth` property value, forcefully cutting off any remaining characters.
- Optionally: If the length of all non-bar characters is computed before each render, then the length of the bar could become dynamic, so that the non-bar data is not truncated. (This could allow for new options `minBarWidth` and `maxBarWidth`.)

Having shared that, I also want to share some feedback about when I first used this module (sometime last year maybe?):

I was confused by the naming of the constructor option `width`. I expected that it would control what I described above as `maxWidth`. I think it would be more clear if the `width` property were renamed `barWidth` in the next major semver release.

**Comment by fuxingZhang** [comment]
### Thank you so much for your feedback.

Since I have turned to golang development for almost 2 years, now I rarely do js and typescript related development, so I am very happy to accept pull requests.

### What is the problem ?

Can't get the tty width now. This is the cause of the bug.
```ts
  private get ttyColumns(): number {
    return 100;
  }
```
https://github.com/deno-library/progress/blob/e5908c3bda93dc710360b7d7ad919e456f29ce89/mod.ts#L197

### Is there a way to get the tty width
Yes! Deno v1.2.0 started to support tty column, but is unstable until now.

in "mod.unstable.ts"
```ts
  private get ttyColumns(): number {
    return Deno.consoleSize(Deno.stdout.rid).columns;
  }
```  
https://github.com/deno-library/progress/blob/e5908c3bda93dc710360b7d7ad919e456f29ce89/mod.unstable.ts#L208

You can try it by following code
```bash
deno run --unstable ./examples/width.unstable.ts
```

### Other   

At the same time, I welcome pull requests. 
`I hope to be compatible with the old code. Destructive updates will cause errors or unexpected results after users update.`

You can add `maxWidth` or `barWidth`, It is great. It would be more clear.
You can rename width to barWidth in the next major semver release. But need clearer documentation.

I don't have enough time and energy to maintain this library.I'm very sorry.

**Comment by fuxingZhang** [comment]
I found the problem.

```ts
if (str.length < this.lastStr.length) {
  str += " ".repeat(this.lastStr.length - str.length);
}
```
There is a bug in the length calculation here, if it is colored.

```ts
complete = bgGreen(" "),
incomplete = bgWhite(" "),
```

The space length is one, but the length is 11 if it is colored.
10 more lengths are calculated.

**Comment by fuxingZhang** [comment]
Landed in [v1.3.0](https://github.com/deno-library/progress/releases/tag/v1.3.0)

**Comment by fuxingZhang** [comment]
Sorry for taking so long to fix this bug
